src/test/tools/pmemobjcli/pmemobjcli.c

```diff
@@ -1733,6 +1907,285 @@ pocli_pmemobj_tx_errno(struct pocli_ctx *ctx, struct pocli_args *args)
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act(ctx, args, 2, &actp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_number(args, 3, &type_num);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_size(args, 4, &size);
+	if (ret)
+		return ret;
+
+	PMEMoid r = pmemobj_reserve(ctx->pop, actp, size, type_num);
+
+	if (oidp != NULL)
+		*oidp = r;
+
+	pocli_printf(ctx, "%s(%s, %s, %zu, %llu): %d\n", args->argv[0],
+		args->argv[1], args->argv[2], size, type_num, r);
+
+	return ret;
+}
+
+/*
+ * pocli_pmemobj_xreserve -- pmemobj_xreserve() command
+ */
+static enum pocli_ret
+pocli_pmemobj_xreserve(struct pocli_ctx *ctx, struct pocli_args *args)
+{
+	if (args->argc != 7)
+		return POCLI_ERR_ARGS;
+
+	PMEMoid *oidp = NULL;
+	struct pobj_action *actp = NULL;
+	uint64_t type_num = 0;
+	uint64_t flags = 0;
+	size_t size = 0;
+	uint64_t class_id = 0;
+	int xalloc_zero = 0;
+	enum pocli_ret ret;
+
+	ret = pocli_args_obj(ctx, args, 1, &oidp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act(ctx, args, 2, &actp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_number(args, 3, &type_num);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_size(args, 4, &size);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_class_id(args, 5, &class_id);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_flag(args, 6, &xalloc_zero);
+	if (ret)
+		return ret;
+
+	if (xalloc_zero != 0)
+		flags = class_id + 1;
+	else
+		flags = class_id;
+
+	PMEMoid r = pmemobj_xreserve(ctx->pop, actp, size, type_num, flags);
+
+	if (oidp != NULL)
+	*oidp = r;
+
+	pocli_printf(ctx, "%s(%s, %s, %zu, %llu, %x): %d\n", args->argv[0],
+		args->argv[1], args->argv[2], size, type_num, flags, r);
+
+	return ret;
+}
+
+/*
+ * pocli_pmemobj_defer_free -- pmemobj_defer_free() command
+ */
+static enum pocli_ret
+pocli_pmemobj_defer_free(struct pocli_ctx *ctx, struct pocli_args *args)
+{
+	if (args->argc != 3)
+		return POCLI_ERR_ARGS;
+
+	PMEMoid *oidp = NULL;
+	struct pobj_action *actp = NULL;
+	enum pocli_ret ret;
+
+	ret = pocli_args_obj(ctx, args, 1, &oidp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act(ctx, args, 2, &actp);
+	if (ret)
+		return ret;
+
+	if (oidp == NULL)
+		return pocli_err(ctx, POCLI_ERR_ARGS, "NULL OID not allowed\n");
+
+	pmemobj_defer_free(ctx->pop, *oidp, actp);
+
+	pocli_printf(ctx, "%s(%s, %s)\n", args->argv[0], args->argv[1],
+		args->argv[2]);
+
+	return ret;
+}
+
+/*
+ * pocli_pmemobj_set_value -- pmemobj_set_value() command
+ */
+static enum pocli_ret
+pocli_pmemobj_set_value(struct pocli_ctx *ctx, struct pocli_args *args)
+{
+	if (args->argc != 5)
+		return POCLI_ERR_ARGS;
+
+	PMEMoid *oidp = NULL;
+	struct pobj_action *actp = NULL;
+	uint64_t *ptr;
+	uint64_t value;
+	enum pocli_ret ret;
+
+	ret = pocli_args_obj(ctx, args, 1, &oidp);
+	if (ret)
+		return ret;
+
+	ret = pocli_args_act(ctx, args, 2, &actp);
+	if (ret)
+		return ret;
+	ret = pocli_args_number(args, 3, &value);
+	if (ret)
+		return ret;
+
+	ptr = (uint64_t *)pmemobj_direct(*oidp);
```

Comment:
This implementation doesn't match what is in pocli_pmemobj_set_value usage string.
I think neither is desirable.

If the 4th argument would be an offset from the beginning of oid (with verification that offset is smaller than size of object - 8) it would much more useful.